Kernel/Bus/USB/xHCI/xHCIController.h

```diff
@@ -0,0 +1,1187 @@
+/*
+ * Copyright (c) 2024, Idan Horowitz <idan.horowitz@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/StdLibExtras.h>
+#include <Kernel/Bus/PCI/Device.h>
+#include <Kernel/Bus/USB/USBController.h>
+#include <Kernel/Bus/USB/xHCI/xHCIRootHub.h>
+#include <Kernel/Memory/TypedMapping.h>
+
+namespace Kernel::USB {
+
+class xHCIInterrupter;
+
+class xHCIController final
+    : public USBController
+    , public PCI::Device {
+public:
+    static ErrorOr<NonnullLockRefPtr<xHCIController>> try_to_initialize(PCI::DeviceIdentifier const& pci_device_identifier);
+    virtual ~xHCIController() override;
+
+    virtual StringView device_name() const override { return "xHCI"sv; }
+
+    virtual ErrorOr<void> initialize() override;
+    virtual ErrorOr<void> reset() override;
+    virtual ErrorOr<void> stop() override;
+    virtual ErrorOr<void> start() override;
+
+    virtual void cancel_async_transfer(NonnullLockRefPtr<Transfer> transfer) override;
+    virtual ErrorOr<size_t> submit_control_transfer(Transfer& transfer) override;
+    virtual ErrorOr<size_t> submit_bulk_transfer(Transfer& transfer) override;
+    virtual ErrorOr<void> submit_async_interrupt_transfer(NonnullLockRefPtr<Transfer> transfer, u16 ms_interval) override;
+
+    virtual ErrorOr<void> initialize_device(USB::Device&) override;
+
+    ErrorOr<void> initialize_endpoint_if_needed(Pipe const&);
+
+    u8 ports() const { return m_ports; }
+
+    ErrorOr<HubStatus> get_port_status(Badge<xHCIRootHub>, u8 port);
+    ErrorOr<void> set_port_feature(Badge<xHCIRootHub>, u8 port, HubFeatureSelector);
+    ErrorOr<void> clear_port_feature(Badge<xHCIRootHub>, u8 port, HubFeatureSelector);
+
+    void handle_interrupt(Badge<xHCIInterrupter>, u16 interrupter_id);
+
+private:
+    xHCIController(PCI::DeviceIdentifier const& pci_device_identifier, Memory::TypedMapping<u8> registers_io_window);
+
+    void take_exclusive_control_from_bios();
+    void intel_quirk_enable_xhci_ports();
+    ErrorOr<void> find_port_max_speeds();
+
+    void event_handling_thread();
+    void hot_plug_thread();
+
+    // Arbitrarily chosen to decrease allocation sizes, can be increased up to 256 if we reach this limit
+    static constexpr size_t max_devices = 64;
+
+    static constexpr PCI::RegisterOffset intel_xhci_usb2_port_routing_offset = static_cast<PCI::RegisterOffset>(0xD0);
+    static constexpr PCI::RegisterOffset intel_xhci_usb2_port_routing_mask_offset = static_cast<PCI::RegisterOffset>(0xD4);
+    static constexpr PCI::RegisterOffset intel_xhci_usb3_port_super_speed_enable_offset = static_cast<PCI::RegisterOffset>(0xD8);
+    static constexpr PCI::RegisterOffset intel_xhci_usb3_port_routing_mask_offset = static_cast<PCI::RegisterOffset>(0xDC);
+
+    struct CapabilityRegisters {
```

Comment:
Also please add spec chapter numbers for these

Reply:
Sure, added.